ActuateTwoDoubleSolenoids

maybenikhil opened the pull request:
@maybenikhil follow this template:

### ActuateTwoDoubleSolenoids:
  **When Activated** - Figures out what position each thing is.
  **While Active** - Flapper and Pusher both come out.
  **When Deactivated** - Returns to original state.

maybenikhil committed:
Code Cleanup and State Conflict Fix

Cleaned up motor factory and made it make more sense instead of (master, slave, slave), i made it (slave, master, slave), because that's electrically how the CAN bus is laid out. Along with that I made an if/then function to fix the funny business that was happening with the fire button pneumatics state function

maybenikhil committed:
cleanup

OCD lol

maybenikhil committed:
added ActuateTwoDoubleSolenoids.java

This should fix the state conflict issues, probably not the cleanest way to do it, as it's like 8 different if/elif/then states, but it's not too horrible.

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,107 @@
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
```
I think we can remove this comment

maybenikhil replied:
done

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,107 @@
+  private final DoubleSolenoid solenoid2;
+	private final Value direction;
+  private final Value endDirection;
+  private int v = 0;
```
`v` tells me nothing.

What this state is supposed to represent? Naming is important for code legibility. Usually you can get more compact as your scope gets deeper.

spacetug replied:
Also the naked `ints`. `0`, `1`, `2`, and `3` give names to make theirs semantic clear, if nothing else it will help keep the code maintainable as changing the value in one place will change everywhere the name is uses.
```java
 static final int CONST_0 = 0;
```
 is how is done in Java.  Make sure to avoid use `CONST_0` if you can find a better name for it.

Anyway an `enum`  looks a better choice to use here. Check them out: [Enums in Java](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html) (but try to avoid overuse it)

maybenikhil replied:
cleaned up, used enum took me way to long to figure out how to work them, still not quite understanding them, but I get the specific use I did.

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,107 @@
+  // Called when another command which requires one or more of the same
+  // subsystems is scheduled to run
+  @Override
+  protected void interrupted() {
```
I have the impression this will not be called on our code.
We don't declare subsystem requirements on any of our commands (as we should).
And maybe that's the whole problem what we have now. Solenoids react a lot slower compared to say the motors and may get in an invalid state because of that.

spacetug replied:
One more thing
`end` and `interrupted` have the same code.
And `end` may always be called `interrupted` or not. Check it first. if not make one of them call the other or create a new function that both call.

maybenikhil replied:
fixed 
``protected void interrupted() {
    end();
  }``

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,107 @@
+  @Override
+  protected void end() {
+    if(v == 0){
+    }
```
The optimizer most likely will remove the branch but cluttering the code is bad (like all these commented lines we have around). Unless you want to keep it here while testing (so you can put a log or something), you should remove it.

spacetug commented on westcoast/src/main/java/frc/motorFactory/MotorFactoryHybrid.java:
```diff
@@ -5,9 +5,9 @@
-        WPI_TalonSRX motorMain = new WPI_TalonSRX(channelMain);
+    public SpeedController create(int channel1, int channelMain, int channel2) {
         WPI_VictorSPX motor1 = new WPI_VictorSPX(channel1);
+        WPI_TalonSRX motorMain = new WPI_TalonSRX(channelMain);
```
I don't think it's a good idea to change the order of the parameters as they are not clearly typed, but mostly is a style issue.
Except for the controllers instantiation. I will keep the same types together and this is just part style, there is another technical reason (icache friendly but it is a long shot, so in the end is just style XD).

Anyway changing in this implementation means you have to change in `MotorFactory` declaration and in each other `MotorFactory` implementation in the project to keep them working (I think we have 3 in total).

maybenikhil replied:
alright, rolled back

spacetug reviewed (commented):
I still need to review the `execute` function but here is the first batch

maybenikhil committed:
Edited to Felipe's Comments

Reverted changes on MotorFactoryHybrid, and reformatted ActuatTwoDoubleSolenoids with enums that were more logical then the integers I was using previously.

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,81 @@
+  private final Value endDirection;
+  public enum state {
+    bothout, flapperout, pusherout, bothin
+  }
```
`Enums` are considered `Types` , and typical naming convention for `Types` in Java follows `PascalCase` style.
Now elements have more style variety, but overall each word in the is somewhat identifiable and they are considered `constants`/`literals` so typically convention is either `PascalCase` or `SCREAMING_SNAKE_CASE` (which is the Java standard).
```suggestion
  private enum State {
    BOTH_OUT, FLAPPER_OUT, PUSHER_OUT, BOTH_IN
  }
```

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,81 @@
+  public enum state {
+    bothout, flapperout, pusherout, bothin
+  }
+  private state fire = state.bothout;
```
```suggestion
  private State fire = State.BOTH_OUT;
```

spacetug commented on westcoast/src/main/java/frc/commands/ActuateTwoDoubleSolenoids.java:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  @Override
+  protected void execute() {
```
This function is called multiple times while active so:
```
s0 = solenoid
s1 = solenoid2
f = endDirection
r = direction

t0 => s0=f - s1=f
t1 => s0=r - s1=f
t1 => s0=r - s1=f
```
and so forth. Until you deactivate:
```
tn-1 => s0=r - s1=f
tn   => s0=f - s1=f
```
This is the `pusher out` scenario. But other scenarios follow a similar logic of going to a specific state (`solenoid` go to `direction` and `solenoid2` go to `endDirection`) when activated and go back to the original state when deactivated.

Is that what you are trying to achieve?

spacetug replied:
Ok `both in` will make `s0=r` and `s1=f` and then `s0=f` and `s1=r`.

I confess I'm a little bit confused.
Can you write what is the expected behavior in the PR description?

maybenikhil replied:
I wrote the expected behavior. As far as I know, the issue was that it would call the other command when interrupted which would make some state functions get angry. However, since I am writing this as a command, I don't know if it will conflict, but if you don't set it up as a "
while held" or have it reset when interrupted or else when the fire button gets called again it must return to the states it wants to be (originally the "both in"). So now it will return to it's uninterrupted position, and stops the command so when it calls it again it goes to the state again. Technically the if/then code isn't needed for the excecute just the end and interrupted, but it still needs to check, the state before and reset it.

spacetug commented:
On `ActivateDoubleSolenoid` the idea of having two `direction` parameters was to establish two deterministic, arbitrary states.

So when activating or deactivate we just force the solenoid to the specific state because that is what is expected without any need of checking.

Remember that there is no external interference and if another system is messing with the state of the solenoid, it is our fault and we should find the error and fix it.

What we may need is some operation on the proper `init` function so we make sure everything is in the correct state when we need them.

maybenikhil committed:
Formatting Fix & Extra Adjustability

Changed the enums to SCREAMING_SNAKE_CASE, and added 2 new value variables in case wires get switched and something becomes invalid.

digthroughtime committed:
made things work

merge it?
